### DISCUSSION TITLE
Transposition has landed

### DISCUSSION THREAD

**Discussion by jcard0na** [cardonabits/haxo-hw#71] [Announcements]
Some of you have asked about transposing capabilities of the haxophone.
I'm happy to announce that @petermoz's implementation for [transposition](https://github.com/cardonabits/haxo-rs/issues/19) just landed in [haxo-rs]().  The gist of it is:

1. Now [`notemap.json`](https://github.com/petermoz/haxo-rs/blob/transpose-mode/notemap.json) is in concert pitch, no longer in `Bb`
2. You can pick your default transposition (i.e. for `Bb`, `Eb`) by changing one parameter in `haxo.service`.  There is no longer the need for a custom `notemap.json` to transpose your instrument.
3. You can change transposition in real-time, by using the new transposition commands.

See https://github.com/cardonabits/haxo-rs/pull/27 and https://github.com/cardonabits/haxo-rs/pull/28 for details.

To use this now, you need to compile `haxo-rs` from the `main` branch on your haxophone.

Stay musical,

**Comment by largerotationlad** [comment]
I'm looking to use this function, but I'm a total rookie on the software side. Does the software image on the Quickstart guide include the transport function? If not, do you have any tips on how I could get this to work on my Haxophone? 
Thanks!

**Reply by jcard0na** [reply]
Hi @largerotationlad,

No, the released images do not contain this feature.  There is some documentation on how to [compile the software on the Pi](https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#compilation-method-1-native-compilation-on-the-raspi) and [how to install the binary](https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#installation-and-launch).  What Raspberry hardware do you use?

**Reply by largerotationlad** [reply]
Thanks for the help! I'm now using a Raspberry Pi 4.

**Reply by largerotationlad** [reply]
Your instructions were great, but my understanding of how to compile stuff is rudimentary. After lots of trial and error, the set of commands below seemed to work with compiling the software and copying it to the SD card--but calling the transpose function (sucking on low B), still does nothing, the haxophone works as normal though.

Can you spot what thing or many things I am doing wrong?

1. Installed USB drive, booted RasPi, ssh to headless RasPi
>> cd /media/usb
>> sudo git clone https://github.com/cardonabits/haxo-rs.git
>> sudo apt install libfluidsynth-dev git libasound2-dev i2c-tools fluid-soundfont-gm
>> cd haxo-rs/

There appears to be a problem with rust permissions when attempting cargo build command.
  error: Permission denied (os error 13) at path "/media/usb/haxo-rs/targete3gKl7"

This command seemed to fix it:
>> sudo chown -R pi .

Then continuing from your instructions:
>> cargo build --release

The compile process takes 4m20s on the RasPi 4
>> systemctl stop haxo

Remove overlay protection
>> sudo raspi-config
Performance Options -> Overlay Filesystem -> Enable Overlay? No

Then copy files
>> cp target/release/haxo001 /usr/local/bin/haxo001
>> cd scripts/systemd
>> sudo ./install.sh

Restore overlay protection
 >> sudo raspi-config
Performance Options -> Overlay Filesystem -> Enable Overlay? Yes
Reboot machine and try low B (no response)

**Reply by jcard0na** [reply]
Hi @largerotationlad 

I don't see anything wrong on the steps that you executed.  These are a few things to check that could help us pinpoint the issue:

1. Does drawing air on `Bb` still work?  (that is for [changing instruments](https://github.com/cardonabits/haxo-hw/blob/main/docs/instructions_safety.md#changing-instrument-sounds)).
 
2. Can you check that the version of the haxophone reported in the logs matches what you get from `git log` in your repository?
```
journalctl -u haxo
(...)
Starting haxophone (version 54fc92a)...
```
3. Can you check that the files on the SD card match the copies on your repo?
```
diff /media/usb/haxo-rs/target/release/haxo001 /usr/local/bin/haxo001
diff /media/usb/haxo-rs/notemap.json /usr/share/haxo/notemap.json
```

Let's start with these.  I'll let you know if some new idea comes to mind...
Cheers,

Javier

**Reply by jcard0na** [reply]
BTW, you should be able to test the binary before installing by running:
```
systemctl stop haxo # just to make sure you don't collide with the installed service
cd /media/usb/haxo-rs
cargo run --release
```
That's faster when doing debugging than having to install the binary every time.

**Reply by largerotationlad** [reply]
Turns out I was doing the overlay protection wrong.  In raspi-config, I needed to turn off overlay protection, reboot the machine, then make the boot drive writable, and reboot the machine.

Then I ran all of your instructions for compiling.  After rebooting, I turned all the overlay and write-protection back on.

After that, success! Thanks for the help

**Reply by largerotationlad** [reply]
I'm going to go blow the minds of the rest of my saxophone section in highschool jazz band lol

**Reply by jcard0na** [reply]
Excellent!  Send a photo of their faces!! 😝

**Reply by largerotationlad** [reply]
The mind-blowing will likely go on Monday morning haha but in the meantime, I hooked the Haxophone up to an OP-1 for a quick little saxophone quartet! 

Its incredible how well the transpose function works, I can't wait to mess around with it some more. 

![signal-2024-02-04-113912_002](https://github.com/cardonabits/haxo-hw/assets/125436990/bd801688-3e53-4420-bbee-8e18e942afc3)

Here's what I hacked together last night at like 2 AM 😅 

https://youtu.be/ioy5V7Am78w

Thanks again for the help getting this setup, I can't wait to see where the Haxophone goes!

**Comment by stardusthamster** [comment]
I just treid out the transpose mechanism. 
Thank you so much for this nice function! It works fine!
